Google.Api.Gax.Grpc/Rest/JsonStateTracker.cs

```diff
@@ -0,0 +1,240 @@
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+
+namespace Google.Api.Gax.Grpc.Rest;
+
+/// <summary>
+/// Simple state tracker to indicate when a server-streamed response is "done".
+/// </summary>
+/// <remarks>
+/// We expect that:
+/// - The overall response is an array
+/// - The later JSON parser will perform fuller validation (e.g. this code
+///   won't spot that [{[{]}}] is broken.
+///   
+/// Note that this is mutable rather than us creating a new instance on each "push"
+/// of a character, because:
+/// - It's not *just* a simple state machine: we count open/close object/arrays
+/// - We could make it a struct containing those counters and a detailed state enum,
+///   it's not obvious that would be easier to use.
+/// </remarks>
+internal sealed class JsonStateTracker
+{
+    private int arrayDepth;
+    private int objectDepth;
+
+    private State state = State.BeforeTopArray;
+
+    /// <summary>
+    /// Pushes a single character, returning the appropriate action to be taken next.
+    /// </summary>
+    public NextAction Push(char c) => state switch
+    {
+        // Errors are unrecoverable.
+        State.Error => NextAction.SignalError,
+        State.BeforeTopArray => PushFromBeforeTopArray(c),
+        State.AfterTopArray => PushFromAfterTopArray(c),
+        State.JustInsideTopArray => PushFromJustInsideTopArray(c),
+        State.WithinObject => PushFromWithinObject(c),
+        State.StringLiteralNoEscape => PushFromStringLiteralNoEscape(c),
+        State.StringLiteralEscape => PushFromStringLiteralAfterEscape(c),
+        _ => throw new InvalidOperationException($"Unknown state {state}")
+    };
+
+    private NextAction PushFromBeforeTopArray(char c)
+    {
+        if (IsJsonWhitespace(c))
+        {
+            return NextAction.IgnoreAndContinue;
+        }
+        if (c == '[')
+        {
+            state = State.JustInsideTopArray;
+            arrayDepth++;
+            return NextAction.IgnoreAndContinue;
+        }
+        state = State.Error;
+        return NextAction.SignalError;
+    }
+
+    private NextAction PushFromAfterTopArray(char c)
+    {
+        if (IsJsonWhitespace(c))
+        {
+            return NextAction.IgnoreAndContinue;
+        }
+        state = State.Error;
+        return NextAction.SignalError;
+    }
+
+    private NextAction PushFromJustInsideTopArray(char c)
+    {
+        switch (c)
+        {
+            case char when IsJsonWhitespace(c):
+            case ',':
+                return NextAction.IgnoreAndContinue;
+            case '{':
+                objectDepth++;
+                state = State.WithinObject;
+                return NextAction.BufferAndContinue;
+            case ']':
+                arrayDepth--;
+                state = State.AfterTopArray;
+                return NextAction.SignalEndOfResponses;
+            default:
```

Comment:
An array of literals is never expected?

Reply:
As the top level array? No, we always expect to get a response *message*.

I guess *strictly* speaking the message could be a FieldMask, Timestamp or Duration which has a direct string representation in JSON instead of a JSON object, but I don't think we need to handle that case. (In reality, all RPCs have Request/Response objects.)